ebu_tt_live/carriage/test/test_filesystem.py

```diff
@@ -48,11 +49,86 @@ def test_emit_document(self):
+        node.provides.return_value = six.text_type
+        fs_carriage = FilesystemProducerImpl(self.test_dir_path)
+        fs_carriage.register_producer_node(node)
+        fs_carriage.emit_data(data, sequence_identifier='testSeq',
+                              sequence_number=1, time_base='clock', clock_mode='local')
+        exported_document_path = os.path.join(self.test_dir_path, 'testSeq_1.xml')
+        assert os.path.exists(exported_document_path)
+        manifest_path = os.path.join(self.test_dir_path, 'manifest_testSeq.txt')
+        assert os.path.exists(manifest_path)
+        assert isinstance(fs_carriage._default_clocks['testSeq'], Clock)
+
+    def test_msg_first_item_missing_availability(self):
+        data = 'live message without availability time'
+        node = MagicMock(spec=IProducerNode)
+        node.provides.return_value = six.text_type
+        fs_carriage = FilesystemProducerImpl(self.test_dir_path)
+        fs_carriage.register_producer_node(node)
+        fs_carriage.emit_data(data, sequence_identifier='testSeq')
+        assert os.listdir(self.test_dir_path) == []
+        assert fs_carriage._default_clocks == {}
+
+    def test_msg_mid_sequence_missing_availability(self):
+        node = MagicMock(spec=IProducerNode)
+        node.provides.return_value = six.text_type
+        fs_carriage = FilesystemProducerImpl(self.test_dir_path)
+        fs_carriage.register_producer_node(node)
+        data = 'document without availability'
+        fs_carriage.emit_data(data, sequence_identifier='testSeq',
+                              sequence_number=1, time_base='clock', clock_mode='local')
+        data = 'live message without availability time'
+        fs_carriage.emit_data(data, sequence_identifier='testSeq')
+        assert len(os.listdir(self.test_dir_path)) == 3  # document, message and manifest
+        exported_document_path = os.path.join(self.test_dir_path, 'testSeq_1.xml')
+        assert os.path.exists(exported_document_path)
+        exported_message_path = os.path.join(self.test_dir_path, 'testSeq_msg_1.xml')
+        assert os.path.exists(exported_message_path)
+        manifest_path = os.path.join(self.test_dir_path, 'manifest_testSeq.txt')
+        assert os.path.exists(manifest_path)
+
+    def test_msg_mid_sequence_partial_missing_availability(self):
+        # This is a quirky edge case that should never really happen
+        # An acceptable workaround could be to take the the last received availability time - local clock
+        # offset for the sequence and compute an estimate using the local clock
+        #
+        # Since this was not a requirement the message gets discarded...
+        #
+        # There are also an unsettling number of odd defined/not defined combinations of arguments at this
+        # level, none of which is specified to trigger a particular response: such as receiving
+        # availability_time but no time_base...etc.
+
+        node = MagicMock(spec=IProducerNode)
```

Comment:
Yes looks like a nasty copypaste... and could indeed be saved on. Since this is a test it is not that paramount. I was admittedly a bit lazy as I started working out these quirky cases...

On the topic of mocking. When it comes to unittesting something 'mocking' means that you replace the outside dependency with some simplified dummy object that would detach you from issues/configuration requirements of other components so that your testing can focus on the unit at hand without noise.

The 1st line defines a node that implements the IProducerNode interface and the 2nd tells it that if its provides function is called it should return six.text_type... This is necessary because otherwise the carriage mechanism would not be compatible with the node when the carriage mechanism does its compatibility checks.